### PR TITLE
Autofix Network Policy

### LINKED ISSUE #156: Autofix for Network Policies
##### ISSUE TYPE
 - [ ] Bug Report
 - [x] Feature Idea

# BUG REPORT

##### SUMMARY

Add autofix support for Network Policy auditing

##### ENVIRONMENT
* Kubeaudit version: -
* Kubeaudit install method: -

##### STEPS TO REPRODUCE

Run 
```
kubeaudit autofix -f  fixtures/namespace_missing_default_deny_netpol.yml
```
##### EXPECTED RESULTS

Add default egress/ingress deny network policy to namespace

##### ACTUAL RESULTS

Nothing is added/removed

# FEATURE IDEA

- [x] If the maintainers agree with the feature as described here, I intend to submit a Pull Request myself.<sup>1</sup>

**Proposal:** Autofix on Network Policy for Namespaces would be present

<sup><small>1</small></sup> <sub>This is the quickest way to get a new feature! We reserve the right to close feature requests, even ones we like, if the proposer does not intend to contribute to the feature and it doesn't fit in our current roadmap.</sub>

### PR DESCRIPTION
**Pull request by nschhina**
##### Description

Add autofix support for Network Policy auditing

For 🎩  try- 
```
kubeaudit autofix -f  fixtures/namespace_missing_default_deny_netpol.yml
```
Add default egress/ingress deny network policy to namespace

#159 to be fixed later

## Fixes (Autofix for Network Policies #156)

closes #156 

##### Type of change

- [x] Bug fix :bug:
- [x] New feature :sparkles:

##### How Has This Been Tested?

- [x]  TestAllResourcesFixV1
- [x] TestExtraResourcesFixV1
- [x] TestExtraResourcesIngressFixV1
- [x]  TestFixV1Beta2
- [x] TestFixV1Beta1
- [x]  TestFixV1

##### Checklist:

- [x] I have :tophat: my changes (A 🎩 specifically includes pulling down changes, setting them up, and manually testing the changed features and potential side effects to make sure nothing is broken)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] New and existing unit tests pass locally with my changes
- [x] The test coverage did not decrease

### CODE REVIEW

File: cmd/autofix.go [hunk 1]
```diff
@@ -57,6 +62,21 @@ func autofix(*cobra.Command, []string) {
 		}
 		toAppend = true
 	}
+	for index := range extraResources {
+		err = writeSingleResourceManifestFile(extraResources[index], extraFile.Name())
+		if err != nil {
+			log.Error(err)
+		}
+		fixedData, err := ioutil.ReadFile(extraFile.Name())
```
Comment (klautcomputing) [inline cmd/autofix.go]: Why are you reading the file here?

File: cmd/autofix_util.go [hunk 2]
```diff
@@ -42,7 +43,10 @@ func fixPotentialSecurityIssue(resource Resource, result Result) Resource {
 		case ErrorSeccompAnnotationMissing, ErrorSeccompDeprecated, ErrorSeccompDeprecatedPod, ErrorSeccompDisabled,
 			ErrorSeccompDisabledPod:
 			resource = fixSeccomp(resource)
+		case ErrorMissingDefaultDenyIngressNetworkPolicy, ErrorMissingDefaultDenyEgressNetworkPolicy, ErrorMissingDefaultDenyIngressAndEgressNetworkPolicy:
+			resource = fixNetworkPolicy(resource, occurrence)
 		}
+
```
Comment (klautcomputing) [inline cmd/autofix_util.go]:
```suggestion
```

File: cmd/k8sruntime_util.go [hunk 3]
```diff
@@ -48,6 +49,23 @@ func setContainers(resource Resource, containers []ContainerV1) Resource {
 	return resource
 }
 
+func setNetworkPolicyFields(nsName string, policyList []string) Resource {
+	obj := &NetworkPolicyV1{}
+	resource := obj.DeepCopyObject()
```
Comment (klautcomputing) [inline cmd/k8sruntime_util.go]: Why are you making the concrete `NetPol` an abstract `runtime.Object`

File: cmd/k8sruntime_util.go [hunk 4]
```diff
@@ -48,6 +49,23 @@ func setContainers(resource Resource, containers []ContainerV1) Resource {
...
+	switch t := resource.(type) {
```
Comment (klautcomputing) [inline cmd/k8sruntime_util.go]: And then you switch over it to get the concrete `NetPol` type again, that doesn't make sense :)

File: cmd/networkPolicies.go [hunk 5]
```diff
@@ -142,7 +147,9 @@ func getNetworkPoliciesResources(namespace string) (netPolList *NetworkPolicyLis
 		for _, resource := range resources {
 			switch kubeType := resource.(type) {
 			case *NetworkPolicyV1:
-				netPolList.Items = append(netPolList.Items, *kubeType)
+				if (*kubeType).ObjectMeta.Namespace == namespace {
```
Comment (klautcomputing) [inline cmd/networkPolicies.go]:
```suggestion
				if ns, ok := (*kubeType).ObjectMeta.Namespace; ok && ns == namespace {
```

File: cmd/k8sruntime_util.go [hunk 6]
```diff
@@ -48,6 +49,23 @@ func setContainers(resource Resource, containers []ContainerV1) Resource {
 	return resource
 }
 
+func setNetworkPolicyFields(nsName string, policyList []string) Resource {
```
Comment (klautcomputing) [inline cmd/k8sruntime_util.go]:
```func setNetworkPolicyFields(nsName string, policyList []string) Resource {
	var np NetworkPolicyV1
	np.Kind = "NetworkPolicy"
	np.APIVersion = "networking.k8s.io/v1"
	np.ObjectMeta.Namespace = nsName
	np.ObjectMeta.Name = "default-deny"
	for _, policy := range policyList {
		np.Spec.PolicyTypes = append(np.Spec.PolicyTypes, networking.PolicyType(policy))
	}
	return np.DeepCopyObject()
}
```

File: cmd/test_util.go [hunk 7]
```diff
@@ -159,6 +159,21 @@ func compareTextFiles(file1, file2 string) bool {
 		return false
 	}
 
+	f1stat, err := f1.Stat()
```
Comment (klautcomputing) [inline cmd/test_util.go]: The problem is this doesn't give you any idea where they diff. The other code gave you a diff, this doesn't anymore.

File: cmd/autofix.go [hunk 8]
```diff
@@ -35,11 +38,16 @@ func autofix(*cobra.Command, []string) {
 		log.Error(err)
 	}
 	defer os.Remove(finalFile.Name())
+	extraFile, err := ioutil.TempFile("", "kubeaudit_autofix_extra")
```
[the remark this answers is not in the document]
Comment (nschhina) [reply cmd/autofix.go]: Forgotten relic. Got rid of it thanks

File: cmd/test_util.go (hunk 7 above)
Comment (nschhina) [reply cmd/test_util.go]: I've moved the stat code to after the scan loop, now we'll get where the files don't match as well as it'd match length of both files at the end. Do you have a better way around this? One thing I thought of was to just set s1 to be the scanner of the larger file, but anyways if files dont match and we scan doesnt output an error but stat does we can imply that files don't match in length hence files missing some content after?

**Comment by nschhina** [comment]
> I have 2 nitpicks and the issue that comments are deleted here in `fixtures/namespace_missing_default_deny_netpol.yml`
> 
> ```
> -    - from:
> -      - namespaceSelector:     # chooses all pods in namespaces labelled with team=operations
> -          matchLabels:
> -            team: operations
> -        podSelector:           # chooses pods with type=monitoring
> -          matchLabels:
> -            type: monitoring
> +  - from:
> +    - namespaceSelector:
> +        matchLabels:
> +          team: operations
> +      podSelector:
> +        matchLabels:
> +          type: monitoring
> ```
> I am fine with fixing this issue in a different PR since it's about comments and not about autofixing NPs

I am aware of this bug, I think it's just that we don't have support for namespaceSelector or from in our comment slices yet, I'll open an issue on this so I don't forget about it

File: cmd/test_util.go (hunk 7 above)
Comment (klautcomputing) [reply cmd/test_util.go]: It's not the cleanest solution, but I think given how complicated a different solution would be this is I guess the best solution.

**Pull request merged**